Source/Core/VideoCommon/TextureCacheBase.cpp

```diff
@@ -293,6 +295,318 @@ void TextureCacheBase::ScaleTextureCacheEntryTo(TextureCacheBase::TCacheEntry* e
+    {
+      for (u32 level = 0; level < config.levels; level++)
+      {
+        u32 level_width = std::max(config.width >> level, 1u);
+        u32 level_height = std::max(config.height >> level, 1u);
+        auto rect = tex->GetConfig().GetMipRect(level);
+        m_readback_texture->CopyFromTexture(tex, rect, layer, level, rect);
+
+        size_t stride = AbstractTexture::CalculateStrideForFormat(config.format, level_width);
+        size_t size = stride * level_height;
+        size_t start = texture_data.size();
+        texture_data.resize(texture_data.size() + size);
+        m_readback_texture->ReadTexels(rect, &texture_data[start], static_cast<u32>(stride));
+      }
+    }
+  }
+  else
+  {
+    PanicAlert("Failed to create staging texture for serialization");
+  }
+
+  p.Do(texture_data);
+}
+
+std::unique_ptr<AbstractTexture> TextureCacheBase::DeserializeTexture(PointerWrap& p)
+{
+  TextureConfig config;
+  p.Do(config);
+
+  std::vector<u8> texture_data;
+  p.Do(texture_data);
+
+  if (p.GetMode() != PointerWrap::MODE_READ)
+    return nullptr;
+
+  std::unique_ptr<AbstractTexture> tex = g_texture_cache->AllocateTexture(config);
+  if (!tex)
+  {
+    PanicAlert("Failed to create texture for deserialization");
+    return nullptr;
+  }
+
+  size_t start = 0;
+  for (u32 layer = 0; layer < config.layers; layer++)
+  {
+    for (u32 level = 0; level < config.levels; level++)
+    {
+      u32 level_width = std::max(config.width >> level, 1u);
+      u32 level_height = std::max(config.height >> level, 1u);
+      size_t stride = AbstractTexture::CalculateStrideForFormat(config.format, level_width);
+      size_t size = stride * level_height;
+      if ((start + size) > texture_data.size())
+      {
+        ERROR_LOG(VIDEO, "Insufficient texture data for layer %u level %u", layer, level);
+        return tex;
+      }
+
+      // TODO: Load multiple layers.
+      if (layer != 0)
+        tex->Load(level, level_width, level_height, level_width, &texture_data[start], size);
+      start += size;
+    }
+  }
+
+  return tex;
+}
+
+void TextureCacheBase::DoState(PointerWrap& p)
+{
+  p.Do(frameCount);
+  p.Do(last_entry_id);
+
+  if (p.GetMode() == PointerWrap::MODE_WRITE || p.GetMode() == PointerWrap::MODE_MEASURE)
+    DoSaveState(p);
+  else
+    DoLoadState(p);
+}
+
+void TextureCacheBase::DoSaveState(PointerWrap& p)
+{
+  std::map<TCacheEntry*, u32> entry_map;
+  auto AddCacheEntryToMap = [&entry_map, &p](TCacheEntry* entry) {
+    auto iter = entry_map.find(entry);
+    if (iter != entry_map.end())
+      return;
+
+    // Skip custom textures since they won't fit in our staging texture.
+    if (entry->is_custom_tex)
+      return;
+
+    u32 id = static_cast<u32>(entry_map.size());
+    entry_map.emplace(entry, id);
+  };
+  auto GetCacheEntryId = [&entry_map](TCacheEntry* entry) -> u32 {
+    auto iter = entry_map.find(entry);
+    return iter != entry_map.end() ? iter->second : static_cast<u32>(-1);
```

Comment:
Why wasn't `0xFFFFFFFFU` just used instead? This also seems like something that should be a symbolic constant somewhere.